File: ios/DebugOverrideMappers.swift
import OctopusCore

/// Converts the JS-facing debug-override payloads (used only by the `debugOverride*` testing
/// hatches) to their native domain-model equivalents. These are development-only affordances,
/// not part of the stable public API surface — the `debugOverride*` functions on `OctopusSDK`
/// are gated behind `@_spi(OctopusInternalTesting)`.
struct DebugOverrideMappers {

  static func toProfileFieldsLock(_ dict: NSDictionary?) -> ProfileFieldsLock? {
    guard let dict = dict as? [String: Any] else { return nil }
    return ProfileFieldsLock(
      nickname: toProfileFieldLockState(dict, "nickname"),
      avatar: toProfileFieldLockState(dict, "avatar"),
      bio: toProfileFieldLockState(dict, "bio")
    )
  }

  private static func toProfileFieldLockState(_ dict: [String: Any], _ key: String) -> ProfileFieldLockState {
    switch dict[key] as? String {
    case "readOnly": return .readOnly
    case "disabled": return .disabled
    default: return .editable
    }
  }

  /// Reads the flat `postEnablePictures` / `postEnablePolls` / `commentEnablePictures` /
  /// `replyEnablePictures` keys the JS wrapper always sends (see `debugOverrideContentOptions.ts`)
  /// — not nested `post` / `comment` / `reply` sub-dictionaries. Each key independently defaults
  /// to `true` (the native default) when absent, matching the JS wrapper's own per-field `?? true`.
  static func toContentOptions(_ dict: NSDictionary?) -> ContentOptions? {
    guard let dict = dict as? [String: Any] else { return nil }
    return ContentOptions(
      post: ContentOptions.PostOptions(
        enablePictures: dict["postEnablePictures"] as? Bool ?? true,
        enablePolls: dict["postEnablePolls"] as? Bool ?? true
      ),
      comment: ContentOptions.CommentOptions(
        enablePictures: dict["commentEnablePictures"] as? Bool ?? true
      ),
      reply: ContentOptions.ReplyOptions(
        enablePictures: dict["replyEnablePictures"] as? Bool ?? true
      )
    )
  }

  static func toTermsAcceptanceMode(_ mode: String?) -> TermsAcceptanceMode? {
    switch mode {
    case "implicit": return .implicit
    case "explicitMultiCheckbox": return .explicitMultiCheckbox
    case "explicitSingleCheckbox": return .explicitSingleCheckbox
    default: return nil
    }
  }
}

File: android/src/main/java/com/octopuscommunity/octopusreactnativesdk/DebugOverrideMappers.kt
package com.octopuscommunity.octopusreactnativesdk

import com.facebook.react.bridge.ReadableMap
import com.octopuscommunity.sdk.domain.model.CommunityConfig
import com.octopuscommunity.sdk.domain.model.ProfileFieldLockState
import com.octopuscommunity.sdk.domain.model.ProfileFieldsLock

/**
 * Converts the JS-facing debug-override payloads (used only by the `debugOverride*` testing
 * hatches) to their native domain-model equivalents. These are development-only affordances, not
 * part of the stable public API surface — see `InternalOctopusApi` on the native side.
 */
object DebugOverrideMappers {

  fun toProfileFieldsLock(map: ReadableMap?): ProfileFieldsLock? {
    if (map == null) return null
    return ProfileFieldsLock(
      nickname = toProfileFieldLockState(map, "nickname"),
      avatar = toProfileFieldLockState(map, "avatar"),
      bio = toProfileFieldLockState(map, "bio"),
    )
  }

  private fun toProfileFieldLockState(map: ReadableMap, key: String): ProfileFieldLockState {
    return when (map.takeIf { it.hasKey(key) }?.getString(key)) {
      "readOnly" -> ProfileFieldLockState.READ_ONLY
      "disabled" -> ProfileFieldLockState.DISABLED
      else -> ProfileFieldLockState.EDITABLE
    }
  }

  /**
   * Reads the flat `postEnablePictures` / `postEnablePolls` / `commentEnablePictures` /
   * `replyEnablePictures` keys the JS wrapper always sends (see `debugOverrideContentOptions.ts`)
   * — not nested `post` / `comment` / `reply` sub-maps. Each key independently defaults to `true`
   * (the native default) when absent, matching the JS wrapper's own per-field `?? true`.
   */
  fun toContentOptions(map: ReadableMap?): CommunityConfig.ContentOptions? {
    if (map == null) return null
    return CommunityConfig.ContentOptions(
      post = CommunityConfig.ContentOptions.PostOptions(
        enablePictures = map.takeIf { it.hasKey("postEnablePictures") }
          ?.getBoolean("postEnablePictures") ?: true,
        enablePolls = map.takeIf { it.hasKey("postEnablePolls") }
          ?.getBoolean("postEnablePolls") ?: true,
      ),
      comment = CommunityConfig.ContentOptions.CommentOptions(
        enablePictures = map.takeIf { it.hasKey("commentEnablePictures") }
          ?.getBoolean("commentEnablePictures") ?: true,
      ),
      reply = CommunityConfig.ContentOptions.ReplyOptions(
        enablePictures = map.takeIf { it.hasKey("replyEnablePictures") }
          ?.getBoolean("replyEnablePictures") ?: true,
      ),
    )
  }

  fun toTermsAcceptanceMode(mode: String?): CommunityConfig.TermsAcceptanceMode? {
    return when (mode) {
      "implicit" -> CommunityConfig.TermsAcceptanceMode.IMPLICIT
      "explicitMultiCheckbox" -> CommunityConfig.TermsAcceptanceMode.EXPLICIT_MULTI_CHECKBOX
      "explicitSingleCheckbox" -> CommunityConfig.TermsAcceptanceMode.EXPLICIT_SINGLE_CHECKBOX
      else -> null
    }
  }
}
